Accelerated inference Blogpost. (#68)
huggingface/blog: accelerated-inference.md

---

> title: "Speeding up inference by 10-100x for 4000 models"

Reviewer: ```suggestion
title: "Speeding up inference by 10-100x for 4000 models"
thumbnail: ./assets/09_accelerated-inference/thumbnail.png
```

Reviewer: add thumbnail here too so you get a nice social sharing thumbnail

---

> # **Speeding up inference by 10-100x for 4000 models**

Reviewer: ```suggestion
# **Speeding up inference by 10-100x for 5,000 models**
```

At HuggingFace we're running a [hosted inference API](https://huggingface.co/pricing) that runs
all the models from the hub for a small price. One of the perks of using the API besides running

> your own is that we provide with accelerated API that speeds up models by up to 100x (mostly likely 10x). This blogpost describes how we are achieving it.

Reviewer: ```suggestion
your own is that we provide an accelerated API that speeds up models by up to 100x (mostly likely 10x). This blogpost describes how we are achieving it.
```

<p align="center">
  <img src="/blog/assets/accelerated-inference/thumbnail.png" />
</p>

This blogpost will assume some knowledge of the internals of transformers in general
and will link to relevant parts for everything we don't explain here.

## Running a pipeline (Open-source and already within transformers)

The first speedup you should get is already implemented for you for free directly
within [transformers](https://github.com/huggingface/transformers). And it's enabled

> by using [pipelines](https://huggingface.co/transformers/main_classes/pipelines.html?highlight=pipelines) instead of coming up with your own code to run the inference.

Reviewer: ```suggestion
by using [pipelines](https://huggingface.co/transformers/main_classes/pipelines.html?highlight=pipelines) instead of rolling out your own code to run the inference.
```

It works only for Seq2Seq models and generation models, but that encompasses the
most popular models that we see used for inference. [gpt2](https://cdn.openai.com/better-language-models/language_models_are_unsupervised_multitask_learners.pdf), [Bart](https://arxiv.org/abs/1910.13461),
[Pegasus](https://arxiv.org/abs/1912.08777), [t5](https://arxiv.org/abs/1910.10683), [gpt3](https://arxiv.org/abs/2005.14165) etc... It's also currently implemented only for [pytorch](https://pytorch.org/)
implementations, but we're working on enabling it on [TF](https://www.tensorflow.org/) too.

Within the architecture of transformers, there are Attention layers which are

> the most important part of it's success. The core of it, is that it's a NxN

Reviewer: ```suggestion
the most important part of its success. The core of it, is that it's a NxN
```

The second part of the speedup is done within the [tokenizers](https://github.com/huggingface/tokenizers/) library.
If you are running `transformers>=4.0` you're most likely using that speedup already !
Tokenization is the process of changing text (a sequence of bytes) into tokens (a sequence
of ints) before feeding it to you preferred model. The tokenization process is usually quite slow
when done within `python` for instance. That's why we created the `tokenizers` library
that is implemented in Rust for performance.

Tokenizing is also an operation that is likely to be cached in most inference situation
so you can also get speedups if you do that correctly. Using `pipelines` from above
will get you all the caching you're going to need.

Overall the Fast tokenizers get a ~10x speedup for the tokenization part, so depending of
your usage it can be a significant part of the speedup too !

## Compiling your model (ONNX, onnxruntime, not fully open source yet)

For the last part of the speedup (another ~10x) we're going to need to dive 
a little bit more into gory details.
We're going to focus on CPU speedups as most inference are run on CPU (it's much cheaper than GPU) but

> the same principles can be also applied to GPU.

Reviewer: ```suggestion
the same principles can also be applied to GPU.
```

---
title: "Speeding up inference by 10-100x for 5000 models"

> thumbnail: ./assets/09_accelerated-inference/thumbnail.png

Reviewer: ```suggestion
thumbnail: /blog/assets/09_accelerated_inference/thumbnail.png
```

---

# **Speeding up inference by 10-100x for 5,000 models**

At HuggingFace we're running a [hosted inference API](https://huggingface.co/pricing) that runs
all the models from the hub for a small price. One of the perks of using the API besides running
your own is that we provide an accelerated API that speeds up models by up to 100x (mostly likely 10x). This blogpost describes how we are achieving it.

<p align="center">
  <img src="/blog/assets/09_accelerated_inference/thumbnail.png" />
</p>

To get a sense of how fast inference becomes, we are getting very close to GPU

> inference speed on a simple CPU (GPU is only 40% faster, but it's 4x times the cost).

Reviewer 2: May be mention than GPU is still very attractive for cases where batching can happen?

|:---------------------------------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------------------:|
|![](/blog/assets/09_accelerated_inference/unoptimized_graph.png)|![](/blog/assets/09_accelerated_inference/optimized_graph.png)|

That's achieved within transformers by using `past_key_values` input, which is
given to you when you call a model. 

Overall this performance boost achieves ~5x improvement.

## Fast tokenizers (Open-source and already within transformers)

The second part of the speedup is done within the [tokenizers](https://github.com/huggingface/tokenizers/) library.
If you are running `transformers>=4.0` you're most likely using that speedup already !
Tokenization is the process of changing text (a sequence of bytes) into tokens (a sequence
of ints) before feeding it to you preferred model. The tokenization process is usually quite slow
when done within `python` for instance. That's why we created the `tokenizers` library

> that is implemented in Rust for performance.

Reviewer 3: Therefore we create a completely new performance-optimized `tokenizers` library implemented in Rust.

> Tokenizing is also an operation that is likely to be cached in most inference situation

Reviewer 3: 2 thats in the sentences again. 

either: an operation "which" is likely.... 
or: speedups of you do "it" correctly...

so you can also get speedups if you do that correctly. Using `pipelines` from above
will get you all the caching you're going to need.

> Overall the Fast tokenizers get a ~10x speedup for the tokenization part, so depending of

Reviewer 3: tokenizers achieves a ~10x speedup

your usage it can be a significant part of the speedup too !

## Compiling your model (ONNX, onnxruntime, not fully open source yet)

> For the last part of the speedup (another ~10x) we're going to need to dive 

Reviewer 3: For the last part of the speedup (another ~10x) we are taking a deep dive into gory details

Please let us know in the comments if you want it !

In order to get the best of this method, you need to be aware that the type of
CPU is important (AVX512 for instance will help tremendously), the fusing of layers
might not work out of the box with the transformers implementations of graphs, so you
need either to edit your ONNX implementations or override some `transformers` methods
in order to get all the inference optimized.
Running advanced `onnxruntime` options is also necessary
like [Gpt2Helper](https://github.com/microsoft/onnxruntime/blob/094384781ea0caa3931061609ca90a84b6a0b64c/onnxruntime/python/tools/transformers/gpt2_helper.py#L100) for instance.
Read the `onnxruntime` documentation thoroughly to get the maximum out of it.
Also be careful when using quantization, there **is** some precision loss, and
can lead to very bad results if not done correctly, so be careful when using it.

Be also careful when trying this method that the actual output of `onnxruntime` 
is platform dependent when you are agressive, meaning that you need to compute
the graph on each machine you are using, you cannot simply move the files around.
 

# Conclusion

> 

Reviewer 3: a calculation with the values would be cool here -> how the 100x are achieved

Overall, we showed how to get ~100x speedup in inference time compared to naïve
implementations, by using *cached* graphs, *native* tokenizers and *compiling*

> your model with a runtime + quantization optimizations. And that the real trick

Reviewer 3: remove "And that"

is to actually combine all those methods at the same time in your product.

> To test all these optimizations right now, you can check out the results in our [Hosted API inference](https://huggingface.co/pricing). Be mindful that we require paid subscription for the accelerated part, but we have a 7-day free trial.

Reviewer 3: i would be a little more aggressive here. 

"To use all these optimizations, you can check out our [Inference API]. To give you a headstart we offer a 7-day free trial. After that, you can choose between 3 different subscription models. 
we not only optimizing all the models but scaling the API according to your needs up to thousands of concurrent requests. 
Besides, it is possible for you to switch between CPU and GPU with a simple parameter change. 
If you want to know more check out our [API product page](Dont have an link here -> pls add)

This trend has introduced daunting challenges for Machine Learning Engineers when deploying the latest models into production. While 100x speedup is a high bar to reach, that’s what it takes to serve predictions with acceptable latency in real-time consumer applications.

To reach that bar, as Machine Learning Engineers at Hugging Face we certainly have an unfair advantage sitting in the same (virtual) offices as the 🤗 Transformers and 🤗 Tokenizers maintainers 😬.  We are also extremely lucky for the rich partnerships we have developed through open source collaborations with hardware and cloud vendors like Intel, NVIDIA, Qualcomm, Amazon and Microsoft that enable us to tune our models x infrastructure with the latest hardware optimizations techniques.

If you want to feel the speed on our infrastructure, start a [free trial](https://huggingface.co/pricing) and we’ll get in touch.

> If you want to benefit from our experience optimizing inference on your own infrastructure, [contact us](mailto:api-enterprise@huggingface.co?subject:Expert%20Acceleration%20Program)! We can support your team through the 🤗 Expert Acceleration Program.

Reviewer 4: Hi @Narsil ,

I think the mailto link should be corrected, otherwise the subject would be empty:
```suggestion
If you want to benefit from our experience optimizing inference on your own infrastructure, [contact us](mailto:api-enterprise@huggingface.co?subject=Expert%20Acceleration%20Program)! We can support your team through the 🤗 Expert Acceleration Program.
```
